pkg/reconciler/namespace/namespace.go

```diff
@@ -137,26 +136,33 @@ func (c *Controller) updateSchedulableCondition(ctx context.Context, ns *corev1.
 // assignCluster attempts to schedule a namespace to a random cluster. If no cluster is
 // available, any existing cluster assignment will be cleared.
 func (c *Controller) assignCluster(ctx context.Context, ns *corev1.Namespace) error {
-	clusters, err := c.clusterLister.List(labels.Everything())
+	allClusters, err := c.clusterLister.List(labels.Everything())
 	if err != nil {
 		return err
 	}
 
-	oldClusterName := ns.Labels[clusterLabel]
+	var clusters []*v1alpha1.Cluster
+	for i := range allClusters {
+		// Only include Clusters that are in the same logical cluster as ns
+		if allClusters[i].ClusterName == ns.ClusterName {
+			clusters = append(clusters, allClusters[i])
+		}
+	}
 
 	// TODO: Filter out un-Ready clusters so a namespace doesn't
 	// get assigned to an un-Ready cluster.
 
 	newClusterName := ""
 	if len(clusters) > 0 {
 		// Select a cluster at random.
-		cluster := clusters[rand.Intn(len(clusters))]
+		cluster := allClusters[rand.Intn(len(allClusters))]
```

Comment:
Is it your intention to filter the set of all clusters down to the set in the same logical cluster as the ns, but then assign from the set of all clusters?